### PR TITLE
Added a TS/RA2 sprite sequence importer for infantry and buildings

### PR DESCRIPTION
**Pull request by Mailaender**
These are traditionally time consuming to do manually.

### CODE REVIEW

File: OpenRA.Mods.TS/UtilityCommands/LegacySequenceImporter.cs [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+				if (!string.IsNullOrEmpty(foundation))
+					ImportBuildingSequences(section);
+			}
+		}
+
+		void ImportBuildingSequences(IniSection section)
+		{
+			Console.WriteLine(section.Name + ":");
+
+			Console.WriteLine("\tDefaults:");
```
Comment (abcdefg30) [inline OpenRA.Mods.TS/UtilityCommands/LegacySequenceImporter.cs]: Won't this create an "empty" `Defaults:` "section" in the yaml, if the imported structure doesn't use tileset extensions?

But anyway, I guess even if, this shouldn't hurt.

**Comment by abcdefg30** [comment]
Hm, seems like this is generation `Facings: 0` values, which is not correct (e.g. import the sequences of the rocketeer and look at its idle definitions) and on other sequences it used `Facings: 6` where it should be 8 (probably not your fault, as the value is really 6, but the assets browser shows me 8 facings...).

GJ otherwise.

**Comment by Mailaender** [comment]
It indeed has some quirks. For example the `stand` sequence always needs to be corrected to:

```diff
	stand:
-		Start: 0
-		Length: 1
-		Facings: 1
	stand:
+		Start: 0
+		Length: 6
+		Facings: 8
```

Those definitions seem to have different meanings for each type although the enumeration looks identical.

File: OpenRA.Mods.TS/UtilityCommands/LegacySequenceImporter.cs (hunk 1 above)
Comment (Mailaender) [inline OpenRA.Mods.TS/UtilityCommands/LegacySequenceImporter.cs]: No tileset codes/extensions is quite the exception for TS/RA2 mods. These auto-generated rules need a cleanup afterwards anyway.

File: OpenRA.Mods.TS/UtilityCommands/LegacySequenceImporter.cs [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+		}
+
+		void ConvertStartLengthFacings(string input)
+		{
+			var splitting = input.Split(',');
+			if (splitting.Length >= 3)
+			{
+				Console.WriteLine("\t\tStart: " + splitting[0]);
+				Console.WriteLine("\t\tLength: " + splitting[1]);
+				Console.WriteLine("\t\tFacings: " + splitting[2]);
```
Comment (GraionDilach) [inline OpenRA.Mods.TS/UtilityCommands/LegacySequenceImporter.cs]: `splitting.Length == 4 ? Console.WriteLine("\t\tFacings: 1") : Console.WriteLine("\t\tFacings: 8")` please.

If I get it right, `splitting[2] - splitting[1]` would be OpenRA `Stride:` but I never seen infantry which either used this logic either used the `Stride` tag in OpenRA.

**Comment by GraionDilach** [comment]
That 0 1 1 stand looks legit though, TS/RA2 couldn't use more frames for idling at all, jumpjet infantry are a giant hack itself, your 0 6 8 fix sounds wrong to me instead tbh. `Length` shouldn't be increased.

**Comment by Mailaender** [comment]
Thanks, added your suggested fixes and some more infantry sequences such as die, panic and cheer.

**Comment by abcdefg30** [comment]
:+1: Looks good to me now.

File: OpenRA.Mods.TS/UtilityCommands/LegacySequenceImporter.cs [hunk 3]
```diff
@@ -0,0 +1,253 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
```
Comment (obrakmann) [inline OpenRA.Mods.TS/UtilityCommands/LegacySequenceImporter.cs]: I can't comment on the functionality, but you should use the updated copyright header here. It's also 2016 :)

**Comment by pchote** [comment]
Any progress here?

**Comment by GraionDilach** [comment]
Technically this doesn't need a progress (excluding the header). I think this is done but haven't found time to check the output to give my thumb.

**Comment by pchote** [comment]
The header is what i'm referring to.

**Comment by Mailaender** [comment]
Updated with the 2016 copyright header.

**Comment by GraionDilach** [comment]
You missed `LegacySequenceImporter`.

**Comment by Mailaender** [comment]
Fixed.

**Comment by pchote** [comment]
You have access to the footprint information, so really should generate the appropriate artwork offsets for buildings (or at least something that is approximately correct).  Ignoring them here and then doing them all by hand (https://github.com/OpenRA/ra2/issues/159) is a giant waste of effort and potential.

**Comment by Mailaender** [comment]
Really, I thought that information just wasn't available and all the TS offsets were corrected manually?

**Comment by pchote** [comment]
It's not available to the OpenRA sprite sequence parsers, because it depends on the building footprints in the rules.  You have access to the building foundation here.

**Comment by GraionDilach** [comment]
For reference:

In TS/RA2 the center of the SHP was the building's top-right (in isometric perspective, the top corner) cells top pixel. OpenRA assumes the center of the SHP is the building's middle.

Besides the footprints, it is based on the actual cell's size, and the offset math is also quite complex at non-square footprints (y offset will always be cellheight*y/2 but the x I'm not certain about, at square footprints, it's 0, but notsure at the rest, might be cellwidth*(x-y)/2 but I'm just guessing).

Including this here might be more complicated than the rest which are universal and shared between TS and RA2 which is why I felt that's being out-of-scope here.

**Comment by GraionDilach** [comment]
Attaching a picture from SHP Builder with TS cellsize to point this out further - although it's known SHP Builder mispositions this grid some pixels away, probably due to rounding errors.

shpbuildertsgrid

**Comment by reaperrr** [comment]
Adding support for automatic building offset generation would be nice.
Needs a rebase in any case.

**Comment by Mailaender** [comment]
Rebased and added the offset guessing, which is indeed an excellent idea and your suggestions work rather good in practice https://github.com/OpenRA/ra2/issues/159 as most houses have a square footprint.

**Comment by pchote** [comment]
The offsets are imported incorrectly for all the rectangular footprints that I checked in TS.

**Comment by Mailaender** [comment]
I don't see any pattern to be honest:

Dimensions: 4,3

```patch
gaweap:
	Defaults:
-		Offset: 0,-36
+		Offset: -12, -42
```

Dimensions: 3,2

```patch
nahand:
	Defaults:
-		Offset: 0,-24
+		Offset: -6, -30
```
	
Dimensions: 2,3
	
```patch
naapwr:
	Defaults:
-		Offset: -24,-36
+		Offset: 12, -30
```

**Comment by pchote** [comment]
Your best test cases will be the TS gates, which are currently imported with an offset that is wrong by multiple cells.

**Comment by Mailaender** [comment]
Dimensions: 3,1

```patch
gagate_a:
	Defaults:
-		Offset: 48,-12
+		Offset: -24, -24
```

Dimensions: 1,3

```patch
gagate_b:
	Defaults:
-		Offset: -48,-36
+		Offset: 24, -24
```

**Comment by Mailaender** [comment]
Removed https://github.com/OpenRA/OpenRA/pull/10715/commits/718107bbc50003398141957655ed6bb1f4e45641, because while this is a good idea, it is a large enough feature for its own pull request and this one already in development for 4 months.

**Comment by pchote** [comment]
From above:
> You have access to the footprint information, so really should generate the appropriate artwork offsets > for buildings (or at least something that is approximately correct). Ignoring them here and then doing
> them all by hand (OpenRA/ra2#159) is a giant waste of effort and potential for this tool.

I don't think it is fair to ask us to merge an unfinished tool upstream just because you don't want to work through the equations to finish it.  That *forces* the burden of effort onto someone else, compared with *asking* for help and receiving it on their timescale.

**Comment by Mailaender** [comment]
In fact I removed the commit to not submit unfinished code, but only proven one.

**Comment by pchote** [comment]
Do you have a branch or commit sha1 for the version with the incorrect offset calculation?

**Comment by pchote** [comment]
`FootprintUtils.CenterOffset` gives you the offset between the center of the top-left cell and the center of the actor.  Just add 512 to Y and then multiply by the tile pixel size / 1024.

**Comment by Mailaender** [comment]
Updated with an improved version of the building offset converter thanks to your help. I just changed the sign and divided the X offset by 2. Updated https://github.com/OpenRA/ra2/pull/224 where the map "Powder Keg" is a good test case, because it features a lot of buildings that are aligned to the street tiles.

File: OpenRA.Mods.TS/UtilityCommands/LegacySequenceImporter.cs [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+		}
+
+		void ImportBuildingSequences(IniSection section)
+		{
+			Console.WriteLine(section.Name + ":");
+
+			Console.WriteLine("\tDefaults:");
+
+			var foundation = section.GetValue("Foundation", string.Empty);
+			if (!string.IsNullOrEmpty(foundation))
```
Comment (pchote) [inline OpenRA.Mods.TS/UtilityCommands/LegacySequenceImporter.cs]: You can simplify this much more:
```csharp
			if (!string.IsNullOrEmpty(foundation))
			{
				var size = foundation.Split('x');
				if (size.Length == 2)
				{
					var x = int.Parse(size[0]);
					var y = int.Parse(size[1]);
	
					var xOffset = (x - y) * grid.TileSize.Width / 4;
					var yOffset = (x + y) * grid.TileSize.Height / 4;
					Console.WriteLine("\t\tOffset: {0},{1}", -xOffset, -yOffset);
				}
			}
```

File: OpenRA.Mods.TS/UtilityCommands/LegacySequenceImporter.cs [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+
+		void ConvertStartLengthFacings(string input)
+		{
+			var splitting = input.Split(',');
+			if (splitting.Length >= 3)
+			{
+				Console.WriteLine("\t\tStart: " + splitting[0]);
+				Console.WriteLine("\t\tLength: " + splitting[1]);
+
+				// TODO splitting[2] - splitting[1] may be Stride:
```
Comment (pchote) [inline OpenRA.Mods.TS/UtilityCommands/LegacySequenceImporter.cs]: Do you have an example of where splitting[2] is used?

Comment (GraionDilach) [inline OpenRA.Mods.TS/UtilityCommands/LegacySequenceImporter.cs]: https://github.com/OpenRA/OpenRA/pull/10715#discussion_r53581402

Comment (GraionDilach) [inline OpenRA.Mods.TS/UtilityCommands/LegacySequenceImporter.cs]: Oh, apparently there are cases in RA2 where this is used (crawl's first frames are reused as prone in RA2). https://github.com/Phrohdoh/oramod-ra2/blob/master/extra-files/art.ini#L9637

Comment (pchote) [inline OpenRA.Mods.TS/UtilityCommands/LegacySequenceImporter.cs]: Ok, then we should resolve this easy TODO before merging.

Comment (Mailaender) [inline OpenRA.Mods.TS/UtilityCommands/LegacySequenceImporter.cs]: I have no idea if I understand what this strides is, but I added it.

File: OpenRA.Mods.TS/UtilityCommands/LegacySequenceImporter.cs (hunk 4 above)
Comment (Mailaender) [inline OpenRA.Mods.TS/UtilityCommands/LegacySequenceImporter.cs]: Added. This won't hide as you commented on a line that isn't affected by your suggestions.

**Comment by pchote** [comment]
This now spits out `Stride: 0` for everything that doesn't need it, and all the non-zero values are negative.

**Comment by pchote** [comment]
For definitions with 3 entries splitting[2] appears to be the stride, and should only be written if it is different to splitting[1].

**Comment by Mailaender** [comment]
Updated the stride calculation with your recent findings.

**Comment by pchote** [comment]
I think i've done enough nit-picking now.  The generated sequences look sensible, so :+1:.

**Comment by pchote** [comment]
This feature isn't shipped, so omitting changelog (and removing the entry for the rules importer).

**Pull request merged**